os/include/tinyara/binfmt/binfmt.h

```diff
@@ -111,6 +111,7 @@ struct binary_s {
 	uint32_t ramstart;		/* Start address of ram partition */
 	uint32_t ramsize;		/* Size of the RAM paritition */
 	uint32_t heapstart;		/* Start address of app heap area */
+	uint32_t datastart;		/* Start address of data section */
```

Comment:
In this case, do we still need `alloc[ALLOC_DATA]`? Because we always keep the datastart address in binary_s.

Reply:
The alloc array is used to hold pointers to allocated sections. At the time of unload / exit, all the pointers kept in alloc array are freed. When reload opt is disabled, we allocate only a single region. So, we can keep only the ALLOC_TEXT in alloc array. But still we need a pointer to the start of data section. So, we use the datastart variable. 

But, we still need ALLOC_DATA when reload opt is enabled. Because, in that case, we actually allocate a separate data section.

Reply:
What is the difference between `datastart` and `ALLOC_DATA`?

Reply:
All the allocated memory regions for the elf are stored in alloc array. And ALLOC_DATA is a an index for holding the pointer for data section in this array. But ALLOC_DATA is valid only if reload opt is enabled and we allocate the data section separately.
But, datastart variable always points to start of data section. Irrespective of whether data section is separately allocated or not.